knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go

```diff
@@ -69,38 +77,119 @@ type ReconcileSourceDeployment struct {
 	// that reads objects from the cache and writes to the apiserver
 	client client.Client
 	scheme *runtime.Scheme
+
+	testShouldUseClusterMonitoring          *bool
+	testShouldGenerateSourceServiceMonitors *bool
 }
 
 // Reconcile reads that state of the cluster for an eventing source deployment
 func (r *ReconcileSourceDeployment) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	reqLogger.Info("Reconciling the source deployment, setting up a service/service monitor if required")
 	dep := &v1.Deployment{}
-	if err := r.client.Get(context.TODO(), request.NamespacedName, dep); err != nil {
-		// the deployment does not exist anymore do nothing
-		return reconcile.Result{}, nil
-	}
-	if err := monitoring.SetupMonitoringRequirements(r.client, dep); err != nil {
+	err := r.client.Get(context.TODO(), request.NamespacedName, dep)
+	inDeletion := false
+	if apierrors.IsNotFound(err) {
```

Comment:
Is 'in deletion' the only state that can cause a 404 ?

Reply:
Afaik this us typical, the first step in reconciliation is to retrieve the object. Either the call fails for some error, object does not exist (due to deletion) or it exists and then you can update its state etc.

Reply:
From below: I think I follow this now. I feel like it's unnecessary to distinguish between "inDeletion" and "disabling monitoring" though. The code paths would likely benefit from handling both cases the same, I believe.